operator/internal/controller/auto/update_controller.go

```diff
@@ -612,6 +634,48 @@ func (s streamReader[T]) Result() (result, error) {
 	return res, fmt.Errorf("didn't receive a result")
 }
 
+// setStatusFromGRPCErr sets the Update object status blocks based on the result of a gRPC error.
```

Comment:
This is the core logic for determining if we have a structured error, and uses it for surfacing information about Pulumi errors back to the user.